vault/identity_store_structs.go

```diff
@@ -0,0 +1,184 @@
+)
+
+const (
+	// The meta key prefix reserved for Vault's internal use
+	metaKeyReservedPrefix = "vault-"
+
+	// The maximum number of metadata key pairs allowed to be registered
+	metaMaxKeyPairs = 64
+
+	// The maximum allowed length of a metadata key
+	metaKeyMaxLength = 128
+
+	// The maximum allowed length of a metadata value
+	metaValueMaxLength = 512
+)
+
+// identityStore is composed of its own storage view and a MemDB which
+// maintains active in-memory replicas of the storage contents indexed by
+// multiple fields.
+type identityStore struct {
+	// identityStore is a secret backend in Vault
+	*framework.Backend
+
+	// view is the storage sub-view where all the artifacts of identity store
+	// gets persisted
+	view logical.Storage
+
+	// db is the in-memory database where the storage artifacts gets replicated
+	// to enable richer queries based on multiple indexes.
+	db *memdb.MemDB
+
+	// validateMountPathFunc is a utility from router which returns the properties
+	// of the mount given the mount path. This is useful to verify the input
+	// supplied over the identity store's API.
+	validateMountPathFunc func(string) *validateMountResponse
+
+	// entityLocks are a set of 256 locks to which all the entities will be
+	// categorized to while performing storage modifications.
+	entityLocks []*locksutil.LockEntry
+
+	// logger is the server logger copied over from core
+	logger log.Logger
+
+	// StoragePacker is used to pack multiple storage entries into 256 buckets
+	storagePacker *storagePacker
+}
+
+// entityStorageEntry represents an entity that gets persisted and indexed.
+// Entity is fundamentally composed of zero or many identities.
+type entityStorageEntry struct {
+	// Identities are the identities that this entity is made of. This can be
+	// empty as well to favor being able to create the entity first and then
+	// incrementally adding identities.
+	Identities []*identityIndexEntry `json:"identities" structs:"identities" mapstructure:"identities"`
+
+	// ID is the unique identifier of the entity which always be a UUID. This
+	// should never be allowed to be updated.
+	ID string `json:"id" structs:"id" mapstructure:"id"`
+
+	// Name is a unique identifier of the entity which is intended to be
+	// human-friendly. The default name might not be human friendly since it
+	// gets suffixed by a UUID, but it can optionally be updated, unlike the ID
+	// field.
+	Name string `json:"name" structs:"name" mapstructure:"name"`
+
+	// Metadata represents the explicit metadata which is set by the clients.
+	// This is useful to tie any information pertaining to the identity. This
+	// is a non-unique field of entity, meaning multiple entities can have the
+	// same metadata set. Entities will be indexed based on this explicit
+	// metadata. This enables virtual groupings of entities based on its
+	// metadata.
+	Metadata map[string]string `json:"metadata" structs:"metadata" mapstructure:"metadata"`
```

Comment:
I wonder if to make it more clear later this should be called `external_metadata` or so. It's not clear that `metadata` as a concept is something the user expects to be derived from users, for instance with tokens although users can set metadata that field is almost always populated by the various auth backends. Another approach would be to actually have this be a struct with two embedded maps, so you have e.g. Metadata.Internal and a Metadata.External.